test/replication-luatest/gh_7253_election_long_wal_write_test.lua

```diff
@@ -0,0 +1,173 @@
+        replication = {
+            server.build_instance_uri('server1'),
+            server.build_instance_uri('server2'),
+            server.build_instance_uri('server3'),
+        },
+    }
+    box_cfg.listen = box_cfg.replication[1]
+    box_cfg.election_mode = 'manual'
+    g.server1 = g.cluster:build_and_add_server({
+        alias = 'server1', box_cfg = box_cfg
+    })
+    box_cfg.listen = box_cfg.replication[2]
+    box_cfg.election_mode = 'voter'
+    g.server2 = g.cluster:build_and_add_server({
+        alias = 'server2', box_cfg = box_cfg
+    })
+    box_cfg.listen = box_cfg.replication[3]
+    g.server3 = g.cluster:build_and_add_server({
+        alias = 'server3', box_cfg = box_cfg
+    })
+    g.cluster:start()
+
+    g.server1:exec(function()
+        local s = box.schema.create_space('test', {is_sync = true})
+        s:create_index('pk')
+    end)
+    wait_fullmesh(g)
+end)
+
+g.after_all(function(g)
+    g.cluster:drop()
+    g.server1 = nil
+    g.server2 = nil
+    g.server3 = nil
+end)
+
+g.test_fence_during_confirm_wal_write = function(g)
+    --
+    -- Make the following topology:
+    --   server2 <-> server1 <- server3
+    --
+    g.server2:exec(function()
+        local replication = table.copy(box.cfg.replication)
+        rawset(_G, 'old_replication', table.copy(replication))
+        replication[3] = nil
+        box.cfg{replication = replication}
+    end)
+    g.server3:exec(function()
+        rawset(_G, 'old_replication', table.copy(box.cfg.replication))
+        box.cfg{replication = {}}
+    end)
+    --
+    -- Server1 starts a txn.
+    --
+    server_block_next_wal_write(g.server2)
+    local f = fiber.new(g.server1.exec, g.server1, function()
+        box.space.test:replace({1})
+    end)
+    f:set_joinable(true)
+    --
+    -- The txn is delivered to server2.
+    --
+    server_wait_wal_is_blocked(g.server2)
+    --
+    -- Server1 receives an ack and blocks on CONFIRM WAL write.
+    --
+    server_block_next_wal_write(g.server1)
+    server_unblock_wal_write(g.server2)
+    server_wait_wal_is_blocked(g.server1)
+    --
+    -- Server3 sends a new term to server1.
+    --
+    local term = g.server1:election_term()
+    fiber.create(g.server3.exec, g.server3, function()
+        box.cfg{
+            election_mode = 'manual',
+            election_timeout = 1000,
+        }
+        -- Silence any exceptions. Only term bump matters and it is checked
+        -- below.
+        pcall(box.ctl.promote)
+    end)
+    g.server1:wait_election_term(term + 1)
+    --
+    -- Server1 finishes CONFIRM WAL write and sees that the synchro queue was
+    -- frozen during the WAL write. Shouldn't affect the result.
+    --
+    server_unblock_wal_write(g.server1)
+    local ok, err = f:join(wait_timeout)
+    t.assert_equals(err, nil)
+    t.assert(ok)
+    t.assert(g.server1:exec(function()
+        return box.space.test:get{1} ~= nil
+    end))
+    --
+    -- Cleanup.
+    --
```

Comment:
Let's maybe put it into `g.after_test` ?
Looks like a good example where g.after_test is actually useful.

Reply:
I don't really like splitting the body of a test case into multiple functions. I wouldn't mind if this would be not related to the test, but in the end I have `wait_fullmesh(g)` which should fail if split-brain happened and this is a part of the test body.

On a broader topic, I don't see any sense in `after_test` for cleaning up something test-specific at all. I suppose it could try to restart instances or restore config if there are many test cases and we really-really want as many of them to run as possible even if one failed. But otherwise it is just a test, simply let it fail. It would work just fine if luatest wouldn't loose the entire .result file if the test hangs even on the last line. I've lost count how many hours I lost on it. As well as on tests having `after_test` and hanging somewhere - they won't produce any .result file either.

Reply:
> It would work just fine if luatest wouldn't loose the entire .result file if the test hangs even on the last line. I've lost count how many hours I lost on it. As well as on tests having after_test and hanging somewhere - they won't produce any .result file either.

I didn't know that `after_test` was the culprit.

Anyway, IMO having prepare/teardown code in before_test/after_test looks more clean. Leave it as is if you like.

Reply:
It is not the culprit if you mean not producing any .result file. Not having `after_test` won't produce a .result file on hang either. The problem is that `after_test` doesn't help to fix it. Thus I simply don't see any sense in it except for doing common cleanup for many test cases at once.

Reply:
>  Not having after_test won't produce a .result file on hang either.

Ok, I see now.